spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java

```diff
@@ -179,64 +190,77 @@ public long executeTask(String taskName, Map<String, String> taskDeploymentPrope
 				.findTaskExecutionInformation(taskName, taskDeploymentProperties, composedTaskRunnerName);
 
+		if (taskExecutionInformation.isComposed()) {
+			handleAccessToken(commandLineArgs, taskExecutionInformation);
+		}
+
 		TaskExecution taskExecution = taskExecutionRepositoryService.createTaskExecution(taskName);
 
-		if (taskExecutionInformation.isComposed()) {
-			boolean containsAccessToken = false;
-			boolean useUserAccessToken = false;
+		AppDeploymentRequest appDeploymentRequest = this.taskAppDeploymentRequestCreator.
+				createRequest(taskExecution, taskExecutionInformation, commandLineArgs, platformName);
 
-			final String dataflowServerAccessTokenKey = "dataflow-server-access-token";
-			final String dataflowServerUseUserAccessToken = "dataflow-server-use-user-access-token";
+		TaskManifest taskManifest = createTaskManifest(platformName, taskExecutionInformation, appDeploymentRequest);
 
-			for (String commandLineArg : commandLineArgs) {
-				if (commandLineArg.startsWith("--" + dataflowServerAccessTokenKey)) {
-					containsAccessToken = true;
-				}
-				if (StringUtils.trimAllWhitespace(commandLineArg).equalsIgnoreCase("--" + dataflowServerUseUserAccessToken + "=true")) {
-					useUserAccessToken = true;
-				}
-			}
+		TaskManifest previousManifest = this.dataflowTaskExecutionMetadataDao.getLastManifest(taskName);
 
-			final String dataflowAccessTokenPropertyKey = "app." + taskExecutionInformation.getTaskDefinition().getRegisteredAppName() + "." + dataflowServerAccessTokenKey;
-			for (Map.Entry<String, String> taskDeploymentProperty : taskExecutionInformation.getTaskDeploymentProperties().entrySet()) {
-				if (taskDeploymentProperty.getKey().equals(dataflowAccessTokenPropertyKey)) {
-					containsAccessToken = true;
-				}
+		if(!taskDeploymentProperties.isEmpty()) {
+			taskDeploymentProperties = appDeploymentRequest.getDeploymentProperties();
+		}
+		else {
+			if(previousManifest != null && !previousManifest.getTaskDeploymentRequest().getDeploymentProperties().equals(taskDeploymentProperties)) {
+				appDeploymentRequest = updateDeploymentProperties(commandLineArgs, platformName, taskExecutionInformation, taskExecution, previousManifest);
+
+				taskDeploymentProperties = previousManifest.getTaskDeploymentRequest().getDeploymentProperties();
 			}
+		}
 
-			if (!containsAccessToken && useUserAccessToken) {
-				final String token = TokenUtils.getAccessToken();
+		AppDeploymentRequest request = new AppDeploymentRequest(appDeploymentRequest.getDefinition(),
+				appDeploymentRequest.getResource(),
+				taskDeploymentProperties,
+				appDeploymentRequest.getCommandlineArguments());
+
+		taskManifest.setTaskDeploymentRequest(request);
+
+		if(!isAppDeploymentSame(previousManifest, taskManifest)) {
+
+			validateAndLockUpgrade(taskName, platformName, taskExecution);
+
+			logger.debug("Deleting %s and all related resources from the platform", taskName);
+			taskLauncher.destroy(taskName);
```

Comment:
Is this too early to delete the task before executing the upgrade? I wonder what happens if the upgrade fails.

Reply:
Too early?  There is only one line of code between the destroy and the launch.  What else should happen before the destroy?

Reply:
I meant the sequence operations, especially when the destroy happens and the upgrade fails after that.

Reply:
There is no "upgrade" after a destroy.  If the launch fails after the destroy, when they try to launch again, it will just attempt to deploy again.  That's how task works normally.  If the app isn't deployed, as part of a launch it will perform the deployment.